database/mysql.go

```diff
@@ -148,77 +153,105 @@ func CreateTestMySQLDB(t *testing.T) *TestMySQLDB {
 		t.Skip("Docker not enabled")
 	}
 
-	config, container, err := RunMySQLDockerInstance(&DatabaseConfig{})
-	if err != nil {
-		t.Fatal(err)
+	mySQLTestDBSetup.Do(func() {
+		var err error
+		sharedMySQLConfig, err = findOrLaunchMySQLContainer()
+		require.NoError(t, err)
+	})
+
+	dbName, err := createTemporaryDatabase(sharedMySQLConfig)
+	require.NoError(t, err)
+
+	dbConfig := &DatabaseConfig{
+		DatabaseName: dbName,
+		MySQL:        sharedMySQLConfig,
 	}
 
 	logger := log.NewNopLogger()
-
 	ctx, cancelFunc := context.WithCancel(context.Background())
-
-	db, err := NewAndMigrate(ctx, logger, *config)
+	db, err := NewAndMigrate(ctx, logger, *dbConfig)
 	if err != nil {
-		container.Close()
 		t.Fatal(err)
 	}
 
 	// Don't allow idle connections so we can verify all are closed at the end of testing
 	db.SetMaxIdleConns(0)
 
 	return &TestMySQLDB{
-		DB:        db,
-		container: container,
-		shutdown:  cancelFunc,
-		t:         t,
+		DB:       db,
+		name:     dbName,
+		shutdown: cancelFunc,
+		t:        t,
 	}
 }
 
-func RunMySQLDockerInstance(config *DatabaseConfig) (*DatabaseConfig, *dockertest.Resource, error) {
-	if config.DatabaseName == "" {
-		config.DatabaseName = "test"
+// We connect as root to MySQL server and create database with random name to
+// run our migrations on it later.
+func createTemporaryDatabase(config *MySQLConfig) (string, error) {
+	dsn := fmt.Sprintf("%s:%s@%s/", "root", config.Password, config.Address)
+	db, err := sql.Open("mysql", dsn)
+	if err != nil {
+		return "", err
 	}
+	defer db.Close()
 
-	if config.MySQL == nil {
-		config.MySQL = &MySQLConfig{}
+	dbName := "test" + base.ID()
+
+	_, err = db.ExecContext(context.Background(), fmt.Sprintf("create database %s", dbName))
+	if err != nil {
+		return "", err
 	}
 
-	if config.MySQL.User == "" {
-		config.MySQL.User = "moov"
+	_, err = db.ExecContext(context.Background(), fmt.Sprintf("grant all on %s.* to '%s'@'%%'", dbName, config.User))
+	if err != nil {
+		return "", err
 	}
 
-	if config.MySQL.Password == "" {
-		config.MySQL.Password = "secret"
+	return dbName, nil
+}
+
+func findOrLaunchMySQLContainer() (*MySQLConfig, error) {
+	var containerName = "mysql-test-container"
```

Comment:
I wonder if we should prefix this with `moov` to avoid conflicting with other projects on a developer's laptop.

Reply:
Agree. Will prefix.